Function signature implemented

ankitpriyarup opened the pull request:
![image](https://media.giphy.com/media/dxPFVzRJlnxMGMe1aF/giphy.gif)
I tried to match up with current code base style as much as I could, let me know about further fixes. I didn't modified much of existing code like before, from now I'll mention what all major changes I have done.
- CompletionItem insert text for functions previously had a '(' appended in the end so for atan it will put atan( upon pressing enter just like how it's there in Godot inbuilt editor. I removed that extra bracket because of two reasons - pressing that bracket character triggers the function signature and if there's a bracket already the signature won't trigger automatically, secondly I've always seen it that way in every code editor always pressing bracket right after typing function seems intuitive, in godot in built editor it makes sense because bracket doesn't self close but with external editor I think this implementation makes more sense. Let me know about your opinions in this one.
- In gdscript_editor.cpp I've added a check for r_options != NULL because I don't want to again push all code completion item while doing signatureHelp because I am not using them
- I wonder what's a good way to implement documentation currently I've simply hard-coded it with a TODO tag
- Comma is also used for triggering function signature, consider the case when the user is editing previous written code then pressing comma should bring the function signature

Geequlim commented:
The idea for ducumentations is to make a symbol pool to cache all of the symbols (both native symbols and script symbols). The server fill the pool at startup by parse all scripts in the workspace for script symbols and load native symbols from the `DocData`.

With this mechanism  we can use `ScriptLanguage::lookup_code` to find symbols under the position we need to provide documentations and signatures from the cache.

ankitpriyarup commented:
I've implemented the documentation feature. This is how it's looking. I've created a map to store these information in simple key value both string format. document string supports markdown maybe we can decorate it more later. 
![image](https://user-images.githubusercontent.com/31309984/59955750-c5c64500-94a9-11e9-9a04-f8d4555db3a7.png)
- I added another parameter to code_complete function i.e. hint_identifier it's in this format - class_name::method_name it is the key to search in the documentation map. Even parameters are stored separately so it's like @GDScript::atan::x for x parameter documentation and similarly for y.

I've also done some tests Implementation seems bug free. Although for type based functions like var a = int(5) there are multiple function signatures possible
![image](https://user-images.githubusercontent.com/31309984/59955966-d925e000-94aa-11e9-9753-e83a14adc172.png)
I will create another last commit to address this functionality rest everything seems fine to me. Hope we can still merge this before GSoC First Evaluation

Geequlim commented on core/script_language.h:
```diff
@@ -200,6 +200,11 @@ class ScriptInstance {
 	virtual ~ScriptInstance();
 };
 
+struct ScriptItemDocumentationType {
```
This type seems only need by the LSP implementation. You'd better avoid declear it in the core part of the engine.

ankitpriyarup replied:
Okay I will move it to gdscript_workspace.h
